src/Microsoft.AspNetCore.SystemWebAdapters.Authentication/RemoteAuthenticationAuthHandler.cs

```diff
@@ -0,0 +1,107 @@
+        return Task.CompletedTask;
+    }
+
+    private async Task<RemoteAuthenticationResult> GetRemoteAuthenticationResultAsync()
+    {
+        if (_remoteAuthenticationResult is null)
+        {
+            // Retrieve the remote authentication result and apply any processors
+            _remoteAuthenticationResult = await _authService.AuthenticateAsync(Context.Request, CancellationToken.None);
+            foreach (var processor in _resultProcessors)
+            {
+                await processor.ProcessAsync(_remoteAuthenticationResult, Context);
+            }
+        }
+
+        return _remoteAuthenticationResult;
+    }
+
+    protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+    {
+        // If remote authentication is enabled for the endpoint, retrieve the remote authentication results and return
+        // an appropriate AuthenticateResult value.
+        if (Context.GetEndpoint()?.Metadata.GetMetadata<IRemoteAuthenticationMetadata>() is IRemoteAuthenticationMetadata metadata && metadata.Enabled)
+        {
+            var authResult = await GetRemoteAuthenticationResultAsync();
+
+            if (authResult.User is not null)
+            {
+                var ticket = new AuthenticationTicket(authResult.User, Scheme.Name);
+                _logger.LogDebug("Authenticated user {UserName} based on remote authentication service", authResult.User.Identity?.Name);
+                return AuthenticateResult.Success(ticket);
+            }
+            else
+            {
+                _logger.LogDebug("Remote service did not authenticate a user");
+            }
+        }
+        else
+        {
+            _logger.LogDebug("Not using remote authentication handler as endpoint metadata does not have remote authentication enabled");
+        }
+
+        return AuthenticateResult.NoResult();
+    }
+
+    protected override async Task HandleChallengeAsync(AuthenticationProperties properties)
+    {
+        // If remote authentication is enabled for the endpoint, attempt to apply challenge behaviors based on
+        // the remote authentication result received from the remote app.
+        if (Context.GetEndpoint()?.Metadata.GetMetadata<IRemoteAuthenticationMetadata>() is IRemoteAuthenticationMetadata metadata && metadata.Enabled)
+        {
+            var authResult = await GetRemoteAuthenticationResultAsync();
+
+            // Propagate headers and status code back to the caller
+            // Different authentication schemes may challenge in different ways in the remote
+            // app, so make a best effort to forward the effects of these challenges by forwarding
+            // configured headers (like Location, perhaps) and status code (like 302 or 401, for example).
+            Context.Response.StatusCode = authResult.StatusCode;
+            foreach (var header in authResult.ResponseHeaders)
+            {
+                Context.Response.Headers.Add(header);
```

Comment:
IHeaderDictionary.Add is a problematic API we're discourage using. E.g. it throws if you have multiple headers with the same name.
```suggestion
                Context.Response.Headers.Append(header.Key, header.Value);
```

Reply:
Fixed